build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/release/TagVersionsTask.java

```diff
@@ -0,0 +1,212 @@
+    public void tagVersion(String version) {
+        this.tagVersion = Version.fromString(version);
+    }
+
+    @Option(option = "output-file", description = "File to output tag information to")
+    public void outputFile(Path file) {
+        this.outputFile = file;
+    }
+
+    @TaskAction
+    public void executeTask() throws IOException {
+        if (tagVersion == null) {
+            throw new IllegalArgumentException("No version to tag specified");
+        }
+
+        LOGGER.lifecycle("Tagging version {} in all version records", tagVersion);
+        var outputFile = this.outputFile;
+        List<String> outputLines = outputFile != null ? new ArrayList<>() : null;
+
+        LOGGER.lifecycle("Adding version record for TransportVersion to [{}]", TRANSPORT_VERSIONS_RECORD);
+        int transportVersionId = processVersionFiles(
+            rootDir.resolve(TRANSPORT_VERSION_FILE_PATH),
+            rootDir.resolve(TRANSPORT_VERSIONS_RECORD)
+        );
+        recordTagInfo(outputLines, TRANSPORT_VERSIONS_RECORD, tagVersion, transportVersionId);
+
+        LOGGER.lifecycle("Adding version record for IndexVersion to [{}]", INDEX_VERSIONS_RECORD);
+        int indexVersionId = processVersionFiles(rootDir.resolve(INDEX_VERSION_FILE_PATH), rootDir.resolve(INDEX_VERSIONS_RECORD));
+        recordTagInfo(outputLines, INDEX_VERSIONS_RECORD, tagVersion, indexVersionId);
+
+        if (outputFile != null) {
+            LOGGER.lifecycle("Writing tag information to [{}]", outputFile);
+            Files.write(outputFile, outputLines, StandardOpenOption.WRITE, StandardOpenOption.TRUNCATE_EXISTING);
+        }
+    }
+
+    private int processVersionFiles(Path javaVersionsFile, Path versionRecordsFile) throws IOException {
+        int versionId = readLatestVersion(javaVersionsFile);
+
+        List<String> versionRecords = Files.readAllLines(versionRecordsFile);
+        if (addVersionRecord(versionRecords, tagVersion, versionId)) {
+            Files.write(versionRecordsFile, versionRecords, StandardOpenOption.WRITE, StandardOpenOption.TRUNCATE_EXISTING);
+        }
+        return versionId;
+    }
+
+    private static int readLatestVersion(Path javaVersionsFile) throws IOException {
+        CompilationUnit java = StaticJavaParser.parse(javaVersionsFile);
+
+        FieldIdExtractor extractor = new FieldIdExtractor();
+        java.walk(FieldDeclaration.class, extractor);   // walks in code file order
+        if (extractor.highestVersionId == null) {
+            throw new IllegalArgumentException("No version ids found in " + javaVersionsFile);
+        }
+        return extractor.highestVersionId;
+    }
+
+    private static class FieldIdExtractor implements Consumer<FieldDeclaration> {
+        private Integer highestVersionId;
+
+        @Override
+        public void accept(FieldDeclaration fieldDeclaration) {
+            var ints = fieldDeclaration.findAll(IntegerLiteralExpr.class);
```

Comment:
This seems fragile, wouldn't any other integer literal get caught up by this?

Reply:
we do only have single int initializers in version classes. And it ignores those that have multiple ints in the initializer.